rest/base_client.go

```diff
@@ -15,71 +15,75 @@ import (
 	"github.com/spf13/viper"
 )
 
-// Use a proxy, "" is none. (This is because the flags package doesn't support Optional)
-// May be a URL, a domain name, or an IP address.  HTTP is assumed as the protocol if
-// none is provided.
-var ProxyAddress string
+type (
+	// Error handling (to call into the telemetry library without
+	// creating a circular dependency.)
+	APIErrorHandler = func(method string, path string, e error)
 
-// Accept a server name other than the expected one in the TLS handshake
-var ExpectedServerName string
-
-// Error handling (to call into the telemetry library without
-// creating a circular dependency.)
-type APIErrorHandler = func(method string, path string, e error)
+	// Authentication handler
+	AuthHandler = func(*http.Request) error
+)
 
 var (
-	apiErrorHandler      APIErrorHandler = nil
-	apiErrorHandlerMutex sync.RWMutex
+	// Use a proxy, "" is none. (This is because the flags package doesn't support Optional)
+	// May be a URL, a domain name, or an IP address.  HTTP is assumed as the protocol if
+	// none is provided.
+	ProxyAddress string
+	// Accept a server name other than the expected one in the TLS handshake
+	ExpectedServerName string
+
+	BaseAPIErrorHandler APIErrorHandler
 )
 
-func reportError(method string, path string, e error) {
-	apiErrorHandlerMutex.RLock()
-	defer apiErrorHandlerMutex.RUnlock()
-	if apiErrorHandler != nil {
-		apiErrorHandler(method, path, e)
+func reportError(client *BaseClient, method string, path string, e error) {
```

Comment:
If we are passing in an instance of `BaseClient` to `reportError`, should `reportError` be another method of `BaseClient`?